[Flake]: CMake sometimes fails to download resources.

I've only seen this happen once, but I figured I'd note it for others that may run into the problem:

When following the recommended build commands in the readme, `cmake --build build-output`, this kicks off the step in the nlohmann json CMake recipe to download the main header file:

https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/8ccccfb8920cd43d62009c2c2d12b970acd11b54/cmake/DownloadNlohmannJson.cmake#L19 

Yesterday, that URL redirected me to [1], and the CURL that CMake was built with apparently was not configured to follow redirects, resulting in the downloaded file being empty (represented by the sha256sum starting with "e3b0c44":

```
[...]
CMake Error at nlohmann_json_project-stamp/nlohmann_json_project-download-Release.cmake:16 (message):
  Command failed: 1

   '/opt/clion-2017.3/bin/cmake/bin/cmake' '-DDEST=<PATH-TO-REPOSITORY>/build-output/external/nlohmann_json/src' '-P' '<PATH-TO-REPOSITORY>/cmake/DownloadNlohmannJson.cmake'

  See also

    <PATH-TO-REPOSITORY>/build-output/external/nlohmann_json/src/nlohmann_json_project-stamp/nlohmann_json_project-download-*.log

$ cat <PATH-TO-REPOSITORY>/build-output/external/nlohmann_json/src/nlohmann_json_project-stamp/nlohmann_json_project-download-err.log 
JSON_URL = https://github.com/nlohmann/json/releases/download/v3.1.2/json.hpp
JSON_SHA256 = fbdfec4b4cf63b3b565d09f87e6c3c183bdd45c5be1864d3fcb338f6f02c1733
DEST = <PATH-TO-REPOSITORY>/build-output/external/nlohmann_json/src
CMake Error at <PATH-TO-REPOSITORY>/cmake/DownloadNlohmannJson.cmake:24 (file):
  file DOWNLOAD HASH mismatch

    for file: [<PATH-TO-REPOSITORY>/build-output/external/nlohmann_json/src/json.hpp]
      expected hash: [fbdfec4b4cf63b3b565d09f87e6c3c183bdd45c5be1864d3fcb338f6f02c1733]
        actual hash: [e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855]
             status: [35;"SSL connect error"]

$ echo -n "" | sha256sum
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
```

Seems like this manifests itself differently depending on your version of CMake, according to [2].  If you're building CMake from scratch, there's an option shown there you can provide at bootstrap time to make CURL-in-CMake follow redirects.


[1] https://github-production-release-asset-2e65be.s3.amazonaws.com/11171548/fde59cb6-27cb-11e8-98f3-11de19cf3b61?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20180622%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20180622T053048Z&X-Amz-Expires=300&X-Amz-Signature=2c620b270cc09a875dc77a449ed3c03efc92370af665bd1a4ebeb27b39bb8da0&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Djson.hpp&response-content-type=application%2Foctet-stream

[2] https://gitlab.kitware.com/cmake/cmake/issues/17240

I am going to treat this like any other build flakiness, here is another time it happened:

https://source.cloud.google.com/results/invocations/f5c04893-a9d9-4802-bbed-1bec26a053f6/targets

```
[ 28%] Building CXX object google/cloud/bigtable/CMakeFiles/bigtable_client.dir/table_config.cc.o
CMake Error at nlohmann_json_project-stamp/nlohmann_json_project-download-Debug.cmake:16 (message):
  Command failed: 1
   '/usr/bin/cmake' '-DDEST=/home/build/external/build-output/external/nlohmann_json/src' '-P' '/home/build/external/cmake/DownloadNlohmannJson.cmake'
  See also
    /home/build/external/build-output/external/nlohmann_json/src/nlohmann_json_project-stamp/nlohmann_json_project-download-*.log
make[2]: *** [external/nlohmann_json/src/nlohmann_json_project-stamp/nlohmann_json_project-download] Error 1
make[1]: *** [google/cloud/storage/CMakeFiles/nlohmann_json_project.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 28%] Building CXX object google/cloud/bigtable/CMakeFiles/bigtable_client.dir/version.cc.o
[ 28%] Linking CXX static library libbigtable_client.a
[ 28%] Built target bigtable_client
make: *** [all] Error 2
The command '/bin/sh -c cmake --build build-output -- -j $(nproc)' returned a non-zero code: 2

```

And another one:

https://source.cloud.google.com/results/invocations/ee9f2266-4494-402a-a3b4-e5b3839a811c/targets

```
CMake Error at nlohmann_json_project-stamp/nlohmann_json_project-download-Debug.cmake:16 (message):
  Command failed: 1
   '/usr/bin/cmake' '-DDEST=/home/build/external/build-output/external/nlohmann_json/src' '-P' '/home/build/external/cmake/DownloadNlohmannJson.cmake'
  See also
    /home/build/external/build-output/external/nlohmann_json/src/nlohmann_json_project-stamp/nlohmann_json_project-download-*.log
make[2]: *** [external/nlohmann_json/src/nlohmann_json_project-stamp/nlohmann_json_project-download] Error 1
```

--- a/cmake/DownloadNlohmannJson.cmake
+++ b/cmake/DownloadNlohmannJson.cmake
@@ -23,34 +23,66 @@ set(JSON_URL
 set(
     JSON_SHA256 63da6d1f22b2a7bb9e4ff7d6b255cf691a161ff49532dcc45d398a53e295835f
     )
-message("JSON_URL = ${JSON_URL}")
-message("JSON_SHA256 = ${JSON_SHA256}")
-message("DEST = ${DEST}")
-file(DOWNLOAD "${JSON_URL}" "${DEST}/json.hpp"
-     EXPECTED_HASH SHA256=${JSON_SHA256})
 
-# Remove the definitions of `operator""_json()` and `operator""_json_pointer`. I
-# know it looks ugly to remove specific lines, but we know the contents of the
-# file exactly (there is a SHA256 hash of it above).
-file(READ "${DEST}/json.hpp" JSON_HPP_CONTENT)
-string(
-    REPLACE
-        [==[
+# Use a function to avoid filling up the global namespace with local variables.
+function (_download_json_hpp)
+    set(sleep_seconds 2)
+    foreach (attempt 1 2 3 4 5)
+        math(EXPR sleep_seconds "2 * ${sleep_seconds}")
+        if (NOT attempt EQUAL 1)
+            message(
+                STATUS
+                    "Will retry after ${sleep_seconds} seconds (attempt #${attempt})."
+                )
+            execute_process(COMMAND "${CMAKE_COMMAND}" -E sleep
+                                    "${sleep_seconds}")
+        endif ()
+        file(DOWNLOAD "${JSON_URL}" "${DEST}/json.hpp" STATUS download_status)
+        list(GET download_status 0 download_error_code)
+        if (download_error_code EQUAL 0)
+            break()
+        endif ()
+    endforeach ()
+
+    if (NOT download_error_code EQUAL 0)
+        message(
+            FATAL_ERROR
+                "Could not download ${JSON_URL} after multiple attempts."
+                " status=${download_status}")
+    endif ()
+
+    file(SHA256 "${DEST}/json.hpp" actual_json_hpp_sha256)
+    if (NOT "${JSON_SHA256}" STREQUAL "${actual_json_hpp_sha256}")
+        message(FATAL_ERROR "Mismatch digest for downloaded json.hpp file."
+                            "\n  expected hash: [${JSON_SHA256}]"
+                            "\n    actual hash: [${actual_json_hpp_sha256}]")
+    endif ()
+    # Remove the definitions of `operator""_json()` and
+    # `operator""_json_pointer`. I know it looks ugly to remove specific lines,
+    # but we know the contents of the file exactly (there is a SHA256 hash of it
+    # above).
+    file(READ "${DEST}/json.hpp" JSON_HPP_CONTENT)
+    string(
+        REPLACE
+            [==[
 inline nlohmann::json operator "" _json(const char* s, std::size_t n)
 {
     return nlohmann::json::parse(s, s + n);
 }
 ]==]
-        "" JSON_HPP_CONTENT "${JSON_HPP_CONTENT}")
-string(REPLACE
-        [==[
+            "" JSON_HPP_CONTENT "${JSON_HPP_CONTENT}")
+    string(REPLACE
+            [==[
 inline nlohmann::json::json_pointer operator "" _json_pointer(const char* s, std::size_t n)
 {
     return nlohmann::json::json_pointer(std::string(s, n));
 }
 ]==]
-        ""
-        JSON_HPP_CONTENT
-        "${JSON_HPP_CONTENT}")
+            ""
+            JSON_HPP_CONTENT
+            "${JSON_HPP_CONTENT}")
 
-file(WRITE "${DEST}/json.hpp" "${JSON_HPP_CONTENT}")
+    file(WRITE "${DEST}/json.hpp" "${JSON_HPP_CONTENT}")
+endfunction ()
+
+_download_json_hpp()

